File: revisit/Insert Interval.go
package leetcode

import "leetcode/util"

// https://leetcode.com/problems/insert-interval/

func insert(intervals [][]int, newInterval []int) [][]int {
	if len(intervals) == 0 {
		return [][]int{newInterval}
	}

	if intervals[0][1] < newInterval[0] {
		return append([][]int{intervals[0]}, insert(intervals[1:], newInterval)...)
	}

	if newInterval[1] < intervals[0][0] {
		return append([][]int{newInterval}, intervals...)
	}

	return insert(intervals[1:], []int{
		util.MinInt(intervals[0][0], newInterval[0]),
		util.MaxInt(intervals[0][1], newInterval[1]),
	})
}

// func insert(intervals [][]int, newInterval []int) [][]int {
// 	output := [][]int{}

// 	if len(intervals) == 0 {
// 		return [][]int{newInterval}
// 	}

// 	if newInterval[1] < intervals[0][0] {
// 		return append([][]int{newInterval}, intervals...)
// 	}

// 	if intervals[len(intervals)-1][1] < newInterval[0] {
// 		return append(intervals, newInterval)
// 	}

// 	for i := 0; i < len(intervals); i++ {
// 		interval := intervals[i]
// 		var next []int = interval

// 		if (interval[0] <= newInterval[0] && newInterval[0] <= interval[1]) ||
// 			(interval[0] <= newInterval[1] && newInterval[1] <= interval[1]) {
// 			next[0] = minInt(intervals[i][0], newInterval[0])
// 			for i < len(intervals) && (intervals[i][1] <= newInterval[1] || intervals[i][0] <= newInterval[1]) {
// 				next[1] = maxInt(intervals[i][1], newInterval[1])
// 				fmt.Println(next)
// 				i++
// 			}
// 			i--
// 		}

// 		output = append(output, next)
// 	}

// 	return output
// }

File: src/python/finished/insert_interval.py
from typing import List

# https://leetcode.com/problems/insert-interval/


class Solution:
    def insert(
        self, intervals: List[List[int]], newInterval: List[int]
    ) -> List[List[int]]:
        s, e = newInterval

        # binary search for the smallest index j such that s <= intervals[j][0]
        idx = len(intervals)
        l, r = 0, len(intervals)
        while l < r:
            m = l + (r - l) // 2
            if intervals[m][0] <= s:
                # l.......m...s..r
                l = m + 1
            else:
                idx = m
                r = m

        res = []

        prev_s, prev_e = intervals[0] if intervals and idx != 0 else newInterval
        for i in range(len(intervals) + 1):
            if i == idx:
                if prev_e >= s:
                    prev_e = max(prev_e, e)
                else:
                    res.append([prev_s, prev_e])
                    prev_s, prev_e = s, e

            if i < len(intervals):
                ss, ee = intervals[i]
                if prev_e >= ss:
                    prev_e = max(prev_e, ee)
                    continue

                res.append([prev_s, prev_e])
                prev_s, prev_e = ss, ee

        res.append([prev_s, prev_e])

        return res
